Initial commit of a bare-bones Level Control Cluster.

JimBuzbee opened the pull request:
Tested as an intensity adjustment cluster with Google Home and Alexa. Work still needs to be done to support more commands and to potentially interact with onOff clusters on the same endpoint.

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,154 @@
+    level: TlvField(0, TlvUInt8),
+    transitionTime: TlvField(1, TlvNullable(TlvUInt16)),
+    optionsMask: TlvField(2, TlvUInt8),
+    optionsOverride: TlvField(3, TlvUInt8),
```
and bitmap like options atribute

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,154 @@
+
+/** @see {@link MatterCoreSpecificationV1_0} § 1.6.6.2 */
+const  MoveCommand = TlvObject({
+    moveMode: TlvField(0, TlvUInt8),
```
movemode is an enum defined in 1.6.6.2

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,154 @@
+const  MoveToLevelCommand = TlvObject({
+    level: TlvField(0, TlvUInt8),
+    transitionTime: TlvField(1, TlvNullable(TlvUInt16)),
+    optionsMask: TlvField(2, TlvUInt8),
```
and options same in all commands

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,154 @@
+
+/** @see {@link MatterCoreSpecificationV1_0} § 1.6.6.3 */
+const  StepCommand = TlvObject({
+    stepMode: TlvField(0, TlvUInt8),
```
it is an enum

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,154 @@
+
+/** @see {@link MatterCoreSpecificationV1_0} § 1.6.6.5  */
+const MoveToClosestFrequencyCommand = TlvObject({
+    frequency: TlvField(0,TlvUInt16),
```
default 0

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,154 @@
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.6 */
+const commands = {
+    /** Move from the current level to the specified level */
+    moveToLevel: OptionalCommand(0x0, MoveToLevelCommand, 0x0, TlvNoResponse), // FIXME - all need resposne?
```
No response is correct that way. Response Y means that a status OK is send out automatically but no specific response

Apollon77 commented on src/matter/cluster/server/LevelControlServer.ts:
```diff
@@ -0,0 +1,35 @@
+
+export const LevelControlClusterHandler: () => ClusterServerHandlers<typeof LevelControlCluster> = () => ({
+    moveToLevel: async ({ request:{level, transitionTime, optionsMask, optionsOverride}, attributes: { currentLevel } }) => { 
+      currentLevel.set(level);
```
Add a todo to implement the optionmask handling as described in 1.6.6.1.1 (or you implent it ... it is a bit of bit fun :-)

JimBuzbee replied:
Will get this PR submitted then will take a look at the bitmap. Although I'd be more comfortable doing bit-twiddling in C rather than Typescript ;-)

Apollon77 commented on src/matter/cluster/server/LevelControlServer.ts:
```diff
@@ -0,0 +1,35 @@
+The moveToLevel commands have been tested in Google Home and in Alexa
+*/
+
+export const LevelControlClusterHandler: () => ClusterServerHandlers<typeof LevelControlCluster> = () => ({
```
Honestly I would implement all functions the same for now ... all with a "set" of the level without more login and add todos to add that later, also add later the onoff but for now set level

JimBuzbee replied:
Have all the functions in, but only the "Move to Level" commands include a target level. The others are along the line of "move current level" with a time limit. Should we throw an "Unimplemented" error on any command that is just stubbed out?

Also have thoughts in the "TODO" comments regarding how much logic should be in the library/command vs. in the caller of the library since the library has no knowledge of the underlying hardware capabilities. Do we need some sort of callback from the library?

Apollon77 reviewed (changes requested):
Thank you for providing it ... I added some comments

JimBuzbee commented:
@Apollon77 - Thanks for the comprehensive review. I think you could go blind reading through those specs ;-) Who knew that adjusting the intensity of of a light bulb could take 12 pages of descriptions and explanations! 

I'll start in on your comments today.

JimBuzbee commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,154 @@
+    maxLevel: OptionalAttribute(0x03, TlvUInt8, {default: 254}),
+
+    /** the frequency at which the device is at CurrentLevel */
+    currentFrequency: Attribute(0x04, TlvUInt16, {default: 0}),
```
Made optional

JimBuzbee commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,154 @@
+};
+
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.6 */
+const commands = {
```
Oops - read wrong column. Added TODO but left optional.

JimBuzbee commented:
Updates pushed

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.6 */
+const commands = {
+    /** Move from the current level to the specified level */
```
nit: Method jsdocs are usually an action starting with a verb at the 3rd person.
Moves from the current level to the specified level.
(starting with a capital letter and ending with a dot)

Same for comments below.

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    startUpCurrentLevel: OptionalAttribute(0x4000, TlvNullable(TlvUInt8),),  
+};
+
+/* TODO - Implement all below and change from Optional to Mandatory */
```
Not sure I understand this point: this file describes the specifications of the cluster.
If the commands below are mandatory, this should be described as mandatory to match the specs.

When you create the implementation, you can stub the unimplemented commands:
step() {
   throw new Error("Not implementation")
}

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.5.8 */
+export const OptionsBitmap = {
+    /** Command execution control - see 1.6.5.8.1. */
```
nit: either use the full link to the specs or don't reference it.
"see 1.6.5.8.1" is not explicit and will be confusing if displayed by itself (by the IDE).
The spec section is already referenced above, so I don't think there is a need to mention it again here.

Same below

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+import { BitFlag, MatterApplicationClusterSpecificationV1_0, TlvBitmap, TlvEnum, TlvField, TlvNullable, TlvObject, TlvUInt16, TlvUInt8 } from "@project-chip/matter.js";
+
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.5.8 */
+export const OptionsBitmap = {
```
Why is this exported? Not sure there is a need to reference this bitmap directly from outside this class.

Why this is not the TlvBitmap object? The name suggests it is, but it is not...
export const OptionsBitmap = TlvBitmap(TlvUInt8, {

Apollon77 replied:
Yes it should be a TlvBitmap

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.5.8 */
+export const OptionsBitmap = {
+    /** Command execution control - see 1.6.5.8.1. */
+    ExecuteIfOff: BitFlag(0), // TODO default: true.
```
This should start be a lowercase letter since this will be the name of the flag when using it.
Bitmap definition will be used to create a flag object:
{
    flag1: boolean,
    flag2: boolean,
}

Same below

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+};
+
+/** @see {@link MatterCoreSpecificationV1_0} § 1.6.6.1 */
+const  MoveToLevelCommand = TlvObject({
```
nit: double space after const -> single space

same below

mfucci replied:
MoveToLevelCommandRequest ?
This is no the definition of the command but only the request or parameters so the name is not accurate.

Same below

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    up = 0x0,
+    down = 0x1,
+};
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.6.3 */
```
nit: new line before the jsodc

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.6.2.1 */
+const enum MoveMode {
+    up = 0x0,
```
nit: enum names are capitalized since they are constants: Up / Down

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+
+/** @see {@link MatterCoreSpecificationV1_0} § 1.6.6.5  */
+const MoveToClosestFrequencyCommand = TlvObject({
+    frequency: TlvField(0, TlvUInt16 ), // FIXME - Default 0, how to set TlvField default?
```
nit: extra space after TlvUInt16

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.4 */
+const features = {
+    /** Dependency with the On/Off cluster */
+    OnOff: BitFlag(0), // TODO default: true.
```
nit: lower first letter

mfucci replied:
I know that this is the name in the specs, but it is really terrible and significantly reduce code readability.
I suggest to use: onOffClusterDependency

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    Lighting: BitFlag(1),
+
+    /** Supports frequency attributes and behavior. */
+    Frequency: BitFlag(2)
```
Same as above, this name is terrible for a flag name.
supportFrequency ?

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    down = 0x1,
+};
+
+ /** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.5 */
```
nit: extra space before "/**"

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+ /** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.5 */
+const attributes = {
+    /** the current level of this device */
+    currentLevel: Attribute(0x0,  TlvNullable(TlvUInt8), {default: null}), 
```
nit: extra space before TlvNullable

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    /** time until the current command is complete in 1/10ths of sec */
+    remainingTime: OptionalAttribute(0x01, TlvUInt16, {default: 0}),
+
+    /** the minimum value of CurrentLevel  */
```
CurrentLevel refers to "currentLevel" above, so a jsdoc @link should be used to generate the correct documentation:
The minimum value of {@link attributes.currentLevel}.

See https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#link

Same below

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+
+/** @see {@link MatterCoreSpecificationV1_0} § 1.6.6.1 */
+const  MoveToLevelCommand = TlvObject({
+    level: TlvField(0, TlvUInt8.bound({min:0, max:254})),
```
This is a TlvUInt8 so the min is already 0 (unsigned int 8 bit), so no need to redeclare that.

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    remainingTime: OptionalAttribute(0x01, TlvUInt16, {default: 0}),
+
+    /** the minimum value of CurrentLevel  */
+    minLevel: OptionalAttribute(0x02, TlvUInt8.bound({ min: 0})),
```
This is a TlvUInt8 so the min is already 0 (unsigned int 8 bit), so no need to redeclare that.

Sam below (all unsigned int are already internally constraints to be >= 0)

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    /** stop moving the current level */
+    stop: OptionalCommand(0x3, StopCommand, 0x3, TlvNoResponse),
+
+    /** same as moveToLevel, but change the on/off status of OnOff device on same endpoint */
```
nit: use @link to moveToLevel

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+ });
+
+ /**
+ * Attributes and commands for Pulse Width Moduation (Provisional)
```
nit: Moduation -> Modulation

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    onOffTransitionTime: OptionalAttribute(0x10, TlvUInt16, {default: 0}),
+
+    /** CurrentLevel when OnOff attr of On/Off cluster on the same endpoint is TRUE */
+    onLevel: WritableAttribute(0x11,  TlvNullable(TlvUInt8), {default: null}),
```
nit: extra space before TlvNullable

mfucci commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,179 @@
+    maxFrequency: OptionalAttribute(0x06, TlvUInt16, {default: 0}),
+
+    /** time to move to/from the target when On/Off received by On/Off cluster on the same endpoint */
+    onOffTransitionTime: OptionalAttribute(0x10, TlvUInt16, {default: 0}),
```
I think this one is also writable

mfucci commented on src/matter/cluster/server/LevelControlServer.ts:
```diff
@@ -0,0 +1,44 @@
+// TODO: Create temporary options based on mask and override. How to expose to user of the library?
+
+export const LevelControlClusterHandler: () => ClusterServerHandlers<typeof LevelControlCluster> = () => ({
+    moveToLevel: async ({ request:{level, transitionTime, optionsMask, optionsOverride}, attributes: { currentLevel } }) => { 
```
nit: no need to reference parameters that you are not using
moveToLevel: async ({ request: { level }, attributes: { currentLevel } }) => {

mfucci reviewed (commented):
I would recommend to only have LevelControlCluster.ts in this MR since it is close to being ready for merge

JimBuzbee commented:
@mfucci - Thanks for the detailed review. Maybe, just maybe, I'll eventually learn Typescript :-) I did take the advice from you and @Apollon77 to start using VS instead of vi. Big step up, but I do miss vi's navigation and search/replace commands :-)

I think I've addressed all of your comments. I left the LevelControlCluysterServer in becuase neither Google Home or Alexa call any of the unimplemented functions, but I did add the recommended "Error" throw if they were to be called.

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,181 @@
+    onLevel: WritableAttribute(0x11, TlvNullable(TlvUInt8), {default: null}),
+
+    /** Time taken to move the current level from the min level to the max level. */
+    onTransitionTime: OptionalWritableAttribute(0x12, TlvNullable(TlvUInt16,), {default: null}),
```
nit, comma too much

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,181 @@
+    defaultMoveRate: OptionalWritableAttribute(0x14, TlvNullable(TlvUInt8)),
+
+    /** Determines the default behavior of some cluster commands. */
+    options: Attribute(0x0f, optionsBitmap),
```
default: 0

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,181 @@
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.5.8 */
+const optionsBitmap = TlvBitmap(TlvUInt8, {
+    /** Dependency on On/Off cluster. */
+    executeIfOff: BitFlag(0), // TODO default: true.
```
where default true is defined?

JimBuzbee replied:
I don't know where I saw that :-) - comment removed

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,181 @@
+    level: TlvField(0, TlvUInt8.bound({max:254})),
+    transitionTime: TlvField(1, TlvNullable(TlvUInt16)),
+    optionsMask: TlvField(2, optionsBitmap),
+    optionsOverride: TlvField(3, optionsBitmap), // TODO: 0 Default - all optionsOverride
```
I think default 0 should work  to set

JimBuzbee replied:
Maybe it's my lack of Typescript knowledge, but I don't see a way to specify a default for "TlvField". Is there a way?

Apollon77 replied:
You are basically right ... there will come a fallback value soon but this is for a different behaviour, so you are right ... :-) leave as comments "// default: ..." please

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,181 @@
+const MoveToLevelCommandRequest = TlvObject({
+    level: TlvField(0, TlvUInt8.bound({max:254})),
+    transitionTime: TlvField(1, TlvNullable(TlvUInt16)),
+    optionsMask: TlvField(2, optionsBitmap),
```
also default 0, also below

Apollon77 replied:
add todo

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,181 @@
+    stopWithOnOff: Command(0x7, StopCommandRequest, 0x7, TlvNoResponse),
+
+    /** Changes current frequency to the requested frequency, or to the closest frequency. */
+    moveToClosestFrequency: OptionalCommand(0x8, MoveToClosestFrequencyCommandRequest, 0x8, TlvNoResponse),
```
make take this command (which is only for the frequencvy cluster) out here and then add it just to the Freq-clister by stating:

```
commands: {
    ....commands,
   
    moveToClosestFrquency: ...
}
```

Apollon77 commented on src/matter/cluster/LevelControlCluster.ts:
```diff
@@ -0,0 +1,181 @@
+};
+
+/** @see {@link MatterApplicationClusterSpecificationV1_0} § 1.6.5 */
+const attributes = {
```
Maybwe rename to commonAttributes and have all in that are not especially LT or FQ and add them then separately ? (see below example for commands)

(Sorry for all these cpmments .. in fact isit is one of the first bigger clusters, so we learn how we want to do it in the future )

JimBuzbee replied:
I _think_ I accomplished this and the "commands" comment by breaking out the frequency specific commands and attributes and then using the "Merge()" function to add them back in before the definition of the "PulseWidthModulationLevelControlCluster". But then again, I'm a Typescript novice so when I push, please evaluate what I did - perhaps there's a better way?

No problem on the large number of comments. This is a learning experience for me and I appreciate the feedback.

Apollon77 replied:
I think "merge2 is not needed ... just do it as I showed in https://github.com/mfucci/node-matter/pull/255#discussion_r1117056833

JimBuzbee replied:
I had trouble trying to incorporate as you commented. Just pushed what I have. Probably still needs work

Apollon77 commented on src/matter/cluster/server/LevelControlServer.ts:
```diff
@@ -0,0 +1,60 @@
+
+import { LevelControlCluster } from "../LevelControlCluster";
+import { ClusterServerHandlers } from "./ClusterServer";
+import { UseOptionalAttributes } from "./ClusterServer";
```
nit, combine with line above

Apollon77 commented:
@JimBuzbee if you like , join discord https://discord.gg/EJysjRsx and we can answer questions maybe more direct then via comments here ...

JimBuzbee commented:
Tried to log into discord, but it insisted on me having an .edu address - no go :-)

Apollon77 commented:
@JimBuzbee Look at https://github.com/mfucci/node-matter/pull/255/commits/1b40bae16cc4f7fe5e2ebc8531f88af8444c96e0 ... I did some changes how I meaned it ... I would do it that way.

JimBuzbee commented:
@Apollon77 - Funny - when you recommended "....commands," I thought you were just using shorthand, not literal "..." Something new for me!

Apollon77 commented:
;-) Added even another commit ... so it is even more to the specs. Will look over all again tomorrow. Too tired

Apollon77 commented:
Ok, should I merge it then or do you have stuff still open?

JimBuzbee commented:
@Apollon77 - Go head and merge. I have an implementation of the PulseWidthModulationLevelControlClusterHandler that is pretty close, but I have no good way to test since nothing uses that cluster.

Apollon77 commented:
Ok, I think we will have that for other clusters too ... so best effort and then we see when someone use it
